File: leetcode2/src/leetcode/sort/QuickSort.java
package leetcode.sort;

import java.util.Arrays;

/**
 * Created by hewking on 2017/3/25.
 */

public class QuickSort {

    public static void main(String[] args){
        int[] arr = new int[]{3,26,42,75,34,65,23};
        quickSort(arr,0,arr.length - 1);
        System.out.print(Arrays.toString(arr));
    }

    private static void quickSort(int[] arr,int start , int end) {
        if (start > end) {
            return;
        }
        int part = part(arr,start, end);
        quickSort(arr,start,part - 1);
        quickSort(arr,part + 1 , end);
    }

    public static int part(int[] arr , int i , int j) {
        int value = arr[i];
        while (i < j) {


            while (i < j && arr[j] >= value){
                j--;
            }
            arr[i] = arr[j];

            while (i < j && arr[i] <= value){
                i++;
            }

            arr[j] = arr[i];

//            swap(arr,i,j);

        }
        arr[i] = value;
        return i;
    }


    public static void swap(int[] arr,int i,int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

}

File: leetcode2/src/leetcode/sort/quicksort.kt
package leetcode.sort

/**
 * Created by test on 2018/1/7.
 */
object quicksort {

    @JvmStatic
    fun main(args : Array<String>) {
        var arr = arrayOf(5,43,73,12,6,86,2334,76,43,23)
        arr.forEach {
            print("${it} ")
        }
        sort(arr,0,arr.size - 1)
        arr.forEach {
            print("${it} ")
        }    }

    fun sort(arr : Array<Int>,start : Int, end : Int) {
        if (start >= end) {
            return
        }

        val index = partition(arr,start,end)
        sort(arr,start,index - 1)
        sort(arr,index + 1 , end)

    }

    private fun partition(arr: Array<Int>, start: Int, end: Int): Int {
        var index = arr[start]
        var i = start
        var j = end
        while (i < j) {
            while (i < j && index <= arr[j]) {
                j --
            }
            arr[i] = arr[j]
            while (i < j && arr[i] <= index ) {
                i ++
            }
            arr[j] = arr[i]
        }
        arr[i] = index
        return i

    }

}